pwiz_tools/Skyline/EditUI/SynchronizedIntegrationDlg.cs

```diff
@@ -134,16 +138,64 @@ private void SelectAlignOption()
 
         private void comboGroupBy_SelectedIndexChanged(object sender, EventArgs e)
         {
-            var newItems = SelectedGroupBy.GetItems(Document, new AnnotationCalculator(Document)).ToArray();
-            if (!ArrayUtil.EqualsDeep(listSync.Items.Cast<object>().ToArray(), newItems))
+            var annotationCalc = new AnnotationCalculator(Document);
+            var newItems = SelectedGroupBy.GetItems(Document, annotationCalc).ToArray();
+            if (ArrayUtil.EqualsDeep(listSync.Items.Cast<object>().ToArray(), newItems))
+                return;
+
+            _idxLastSelected = -1;
+
+            listSync.Items.Clear();
+            listSync.Items.AddRange(newItems);
+
+            HashSet<ChromatogramSet> selectedChroms;
+            if (string.IsNullOrEmpty(_memory.Item1))
             {
-                var allChecked = IsAll;
-                listSync.Items.Clear();
-                listSync.Items.AddRange(newItems);
-                cbSelectAll.Checked = false;
-                for (var i = 0; i < listSync.Items.Count; i++)
-                    listSync.SetItemChecked(i, allChecked);
+                // replicates
+                selectedChroms = Document.MeasuredResults.Chromatograms
+                    .Where(chromSet => _memory.Item2.Contains(chromSet.Name)).ToHashSet();
             }
+            else
+            {
+                // annotation
+                var replicateValue = ReplicateValue.FromPersistedString(Document.Settings, _memory.Item1);
+                selectedChroms = Document.MeasuredResults.Chromatograms
+                    .Where(chromSet => _memory.Item2.Contains(replicateValue.GetValue(annotationCalc, chromSet))).ToHashSet();
+            }
+
+            var toCheck = new HashSet<string>();
+            if (string.IsNullOrEmpty(SelectedGroupBy.PersistedString))
+            {
+                // replicates
+                toCheck = selectedChroms.Select(chromSet => chromSet.Name).ToHashSet();
+            }
+            else
+            {
+                // annotation
+                foreach (var item in newItems)
+                {
+                    var thisChroms = Document.MeasuredResults.Chromatograms.Where(chromSet =>
+                        Equals(item, SelectedGroupBy.ReplicateValue.GetValue(annotationCalc, chromSet))).ToArray();
+                    var selectCount = thisChroms.Count(chromSet => selectedChroms.Contains(chromSet));
+                    if (selectCount == thisChroms.Length)
+                    {
+                        toCheck.Add(item.ToString());
+                    }
+                    else if (selectCount != 0)
+                    {
+                        toCheck.Clear();
+                        break;
+                    }
+                }
+            }
+
+            listSync.ItemCheck -= listSync_ItemCheck;
+
+            cbSelectAll.CheckedChanged -= cbSelectAll_CheckedChanged;
```

Comment:
I would recommend that "SetCheckedItems" sets a field called "_changingCheckedItems" and then make it so that both "cbSelectAll_CheckedChanged" and "listSync.ItemCheck" check whether "_changingCheckedItems" is true and if so, return immediately.

This code to remove event handlers and then add them back after you are done is difficult to maintain.